typing/typecore.ml

```diff
@@ -4417,15 +4358,19 @@ and type_label_exp create env loc ty_expected
     with exn when maybe_expansive arg -> try
```

Comment:
The logic here is very tricky, and I'm not fully satisfied with the proposal in the PR:

1. it leaves some logic out of the wrapper that clearly belongs to a `post`, and
2. it performs the snapshotting in a way that does not respect the wrapper nesting.

Sketch of alternative proposal (untested):

```ocaml
and type_label_exp create env loc ty_expected
          (lid, label, sarg) =
  (* Here also ty_expected may be at generic_level *)
  let separate = !Clflags.principal || Env.has_local_constraints env in
  let (vars, ty_arg) =
    with_local_level begin fun () ->
      let (vars, ty_arg) =
        with_local_level_iter_if separate begin fun () ->
          let (vars, ty_arg, ty_res) =
            with_local_level_iter_if separate ~post:generalize_structure
              begin fun () ->
                let ((_, ty_arg, ty_res) as r) = instance_label true label in
                (r, [ty_arg; ty_res])
              end
          in
          begin try
            unify env (instance ty_res) (instance ty_expected)
          with Unify err ->
            raise (Error(lid.loc, env, Label_mismatch(lid.txt, err)))
          end;
          (* Instantiate so that we can generalize internal nodes *)
          let ty_arg = instance ty_arg in
          ((vars, ty_arg), [ty_arg])
        end
        ~post:generalize_structure
      in

      if label.lbl_private = Private then
        if create then
          raise (Error(loc, env, Private_type ty_expected))
        else
          raise (Error(lid.loc, env, Private_label(lid.txt, ty_expected)));

      let arg = if vars = [] then Either.Left 

      (vars, ty_arg)
    end
  in
  let arg =
    let exp_instance arg =
      { arg with exp_type = instance arg.exp_type } in
    let exception Retry of exn in
    try
      let snap = if vars = [] then None else Some (Btype.snapshot ()) in
      exp_instance @@ with_local_level
        begin fun () ->   
          type_argument env sarg ty_arg (instance ty_arg)
        end
        ~post:begin fun arg ->
          if vars <> [] then try
            if maybe_expansive arg then
              lower_contravariant env arg.exp_type;
            generalize_and_check_univars env "field value" arg label.lbl_arg vars;
          with exn when maybe_expansive arg ->
            (* Try to retype without propagating ty_arg, cf PR#4862 *)
            Option.iter Btype.backtrack snap;
            raise (Retry exn)
        end
    with Retry exn -> try
      let arg = with_local_level (fun () -> type_exp env sarg)
          ~post:(fun arg -> lower_contravariant env arg.exp_type)
      in
      exp_instance @@ with_local_level
        begin fun () ->
          let arg = exp_instance arg in
          unify_exp env arg (instance ty_arg);
          arg
        end
        ~post: begin fun arg ->
          generalize_and_check_univars env "field value" arg label.lbl_arg vars
        end
    with Error (_, _, Less_general _) as e -> raise e
    | _ -> raise exn    (* In case of failure return the first error *)
  in
  (lid, label, arg)
```

Reply:
I'm not too excited by this change just now: while I agree that this makes sense, the code is much changed, and the need to introduce a local exception to be sure that we are not catching an exception coming from `type_argument` is hairy.
I don't think we really gain in complexity as a result.
It seems that you are trying to fix smells in the original code rather than in this PR, which tries to be minimal.
I agree that the structure of the code is not so nice, but can can we also differ this one for later ?
To answer your points:
1. what belongs to `post` is not always easy to determine; this is is a first attempt
2. since snapshotting is only about unification, it is not directly related to levels (but I agree that your proposal looks better ultimately)

Reply:
See #11900 for a proposal to merge this change now that the current PR is merged.